pilot/pkg/model/network.go

```diff
@@ -202,3 +326,239 @@ func gcd(x, y int) int {
+	c, err := newClient()
+	if err != nil {
+		return nil, err
+	}
+	return newNetworkGatewayNameCacheWithClient(c), nil
+}
+
+// newNetworkGatewayNameCacheWithClient exported for test
+func newNetworkGatewayNameCacheWithClient(c *dnsClient) *networkGatewayNameCache {
+	return &networkGatewayNameCache{client: c, cache: map[string]nameCacheEntry{}}
+}
+
+// Resolve takes a list of hostnames and returns a map of names to addresses
+func (n *networkGatewayNameCache) Resolve(names sets.Set) map[string][]string {
+	n.Lock()
+	defer n.Unlock()
+
+	n.cleanupWatches(names)
+
+	out := make(map[string][]string, len(names))
+	for name := range names {
+		out[name] = n.resolveFromCache(name)
+	}
+
+	return out
+}
+
+// cleanupWatches cancels any scheduled re-resolve for names we no longer care about
+func (n *networkGatewayNameCache) cleanupWatches(names sets.Set) {
+	for name, entry := range n.cache {
+		if names.Contains(name) {
+			continue
+		}
+		entry.timer.Stop()
+		delete(n.cache, name)
+	}
+}
+
+func (n *networkGatewayNameCache) resolveFromCache(name string) []string {
+	if entry, ok := n.cache[name]; ok && entry.expiry.After(time.Now()) {
+		return entry.value
+	}
+	// ideally this will not happen more than once for each name and the cache auto-updates in the background
+	// even if it does, this happens on the SotW ingestion path (kube or meshnetworks changes) and not xds push path.
+	return n.resolveAndCache(name)
+}
+
+func (n *networkGatewayNameCache) resolveAndCache(name string) []string {
+	if entry, ok := n.cache[name]; ok {
+		entry.timer.Stop()
+	}
+	delete(n.cache, name)
+	addrs, ttl := n.resolve(name)
+	// avoid excessive pushes due to small TTL
+	if ttl < MinGatewayTTL {
+		ttl = MinGatewayTTL
+	}
+	expiry := time.Now().Add(ttl)
+	n.cache[name] = nameCacheEntry{
+		value:  addrs,
+		expiry: expiry,
+		// TTL expires, try to refresh TODO should this be < ttl?
+		timer: time.AfterFunc(ttl, n.refreshAndNotify(name)),
+	}
+
+	return addrs
+}
+
+// refreshAndNotify is triggered via time.AfterFunc and will recursively schedule itself that way until timer is cleaned
+// up via cleanupWatches.
+func (n *networkGatewayNameCache) refreshAndNotify(name string) func() {
+	return func() {
+		log.Debugf("network gateways: refreshing DNS for %s", name)
+		n.Lock()
+		old := n.cache[name]
+		addrs := n.resolveAndCache(name)
+		n.Unlock()
+
+		if !stringSliceEqual(old.value, addrs) {
+			log.Debugf("network gateways: DNS for %s changed: %v -> %v", name, old.value, addrs)
+			n.NotifyGatewayHandlers()
+		}
+	}
+}
+
+// avoid import cycle
+func stringSliceEqual(a, b []string) bool {
+	if len(a) != len(b) {
+		return false
+	}
+
+	for i := range a {
+		if a[i] != b[i] {
+			return false
+		}
+	}
+
+	return true
+}
+
+// resolve gets all the A and AAAA records for the given name
+func (n *networkGatewayNameCache) resolve(name string) ([]string, time.Duration) {
+	// TODO figure out how to query only A + AAAA
```

Comment:
I think you need to do two queries.

Reply:
Various go programs likely do this if we want to peek at their code. Coredns, tailscale,